packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1579,34 +1579,115 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     late bool result;
     assert(() {
       if (_debugDisposed) {
-        result = false;
-        return true;
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('A disposed RenderObject was mutated.'),
+          DiagnosticsProperty<RenderObject>(
+            'The disposed RenderObject was',
+            this,
+            style: DiagnosticsTreeStyle.errorProperty,
+          ),
+        ]);
       }
-      if (owner != null && !owner!.debugDoingLayout) {
+
+      final PipelineOwner? owner = this.owner;
+      // Detached nodes are allowed to mutate and the "can perform mutations`
+      // check will be performed when they re-attach. This assert is
+      // only useful during layout.
+      if (owner == null || !owner.debugDoingLayout) {
         result = true;
```

Comment:
Only when the result is true I think (so the `assert` doesn't assert). One thing this provides is since `result` is a `late final`, if `_debugCanPerformMutations` is accidentally called in release mode it crashes.